distributed/client.py

```diff
@@ -532,13 +532,7 @@ def asynchronous(self):
         though the client was originally created in asynchronous mode we may
         find ourselves in contexts when it is better to operate synchronously.
         """
-        result = self._asynchronous
-        try:
-            if self.loop._thread_ident and get_thread_identity() != self.loop._thread_ident:
```

Comment:
Could this also be handled by the LoopRunner?  We could capture the thread of the IOLoop at the same time we wait for the `threading.Event`

Reply:
That only works if the loop wasn't started by the user.

Reply:
Also, the same pattern occurs in `worker.py` below which doesn't have a LoopRunner.

Reply:
This code is useful because it allows us to use clients that were created to work asynchronously safely outside of an asynchronous environment.

```python
@gen.coroutine
def f():
    client = yield Client(asynchronous=True)
    ...

client = ... get the same client ...
client.gather(...)  # this identifies that it is outside of the IOLoop and uses the blocking API
```

There are likely cleaner ways around this, such as by creating multiple clients or asking users to explicitly annotate all calls with `asynchronous=True/False`.  I'm -1 on the second option (users don't understand this).  The first option is possible, but would require some work on our part.

Reply:
I suspect my replacement code below should work, since `self.loop is IOLoop.current()` is false if you call it from another thread.

Reply:
OK, you're right, that's probably nicer regardless.  Thanks!